sgkit_bgen/bgen_reader.py

```diff
@@ -0,0 +1,186 @@
+            self.pos = variant_arrs["pos"]
+
+            def split_alleles(alleles, block_info=None):
+                if block_info is None or len(block_info) == 0:
+                    return alleles
+
+                def split(allele_row):
+                    return np.array(allele_row[0].split(",")[:2], dtype="S1")
+
+                return np.apply_along_axis(split, 1, alleles[:, np.newaxis])
+
+            self.variant_alleles = variant_arrs["allele_ids"].map_blocks(split_alleles)
+
+        with bgen_file(self.path) as bgen:
+            sample_path = self.path.with_suffix(".sample")
+            if sample_path.exists():
+                self.sample_id = read_samples_file(sample_path, verbose=False)
+            else:
+                if bgen.contain_samples:
+                    self.sample_id = bgen.read_samples()
+                else:
+                    self.sample_id = generate_samples(bgen.nsamples)
+
+        self.shape = (self.n_variants, len(self.sample_id))
+        self.dtype = dtype
+        self.ndim = 2
+
+    def __getitem__(self, idx):
+        if not isinstance(idx, tuple):
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be tuple (received {type(idx)})"
+            )
+        if len(idx) != self.ndim:
+            raise IndexError(  # pragma: no cover
+                f"Indexer must be two-item tuple (received {len(idx)} slices)"
+            )
+
+        if idx[0].start == idx[0].stop:
+            return np.empty((0, 0), dtype=self.dtype)
+
+        start_partition = idx[0].start // self.partition_size
+        start_partition_offset = idx[0].start % self.partition_size
+        end_partition = (idx[0].stop - 1) // self.partition_size
+        end_partition_offset = (idx[0].stop - 1) % self.partition_size
+
+        all_vaddr = []
+        with bgen_metafile(self.metafile_filepath) as mf:
+            for i in range(start_partition, end_partition + 1):
+                partition = mf.read_partition(i)
+                start_offset = start_partition_offset if i == start_partition else 0
+                end_offset = (
+                    end_partition_offset + 1
+                    if i == end_partition
+                    else self.partition_size
+                )
+                vaddr = partition["vaddr"].tolist()
+                all_vaddr.extend(vaddr[start_offset:end_offset])
+
+        with bgen_file(self.path) as bgen:
+            genotypes = [bgen.read_genotype(vaddr) for vaddr in all_vaddr]
+            all_probs = [genotype["probs"] for genotype in genotypes]
+            d = [(2 * probs[:, -1] + probs[:, 1]) for probs in all_probs]
+            return np.stack(d)[:, idx[1]]
+
+
+def read_bgen(
+    path: PathType,
+    chunks: Union[str, int, tuple] = "auto",
+    lock: bool = False,
+    persist: bool = True,
+) -> Dataset:
+    """Read BGEN dataset
+
+    Loads a single BGEN dataset as dask arrays within a Dataset
+    from a bgen file.
+
```

Comment:
Doc bi-allelic constraint.  And probably diploid constraint too?

Reply:
Done in d41914ffc62e75b8a067bd10cdc3907e5af0f0d7